### PR TITLE
feat: read groth parameters lazily via mmap (coarsely)

### PR DESCRIPTION
**Pull request by cryptonemo**

### CODE REVIEW

**Comment by cryptonemo** [comment]
Requires: https://github.com/filecoin-project/bellman/pull/37

**Comment by ValarDragon** [comment]
out of curiousity, do you know how much of a speed improvement this is?

**Comment by cryptonemo** [comment]
> out of curiousity, do you know how much of a speed improvement this is?

Thanks for asking, but the purpose of this PR is strictly runtime RAM reduction, not necessarily speedup type improvements.  In my testing, runtime speed is about the same.

**Comment by ValarDragon** [comment]
Thanks for the explanation!

**Comment by dignifiedquire** [comment]
@cryptonemo I think you have to update the rust version to make the tests work

File: filecoin-proofs/src/caches.rs [hunk 1]
```diff
@@ -62,14 +62,6 @@ where
     Ok(res)
 }
 
-#[inline]
-pub fn lookup_groth_params<F>(identifier: String, generator: F) -> Result<Arc<Bls12GrothParams>>
-where
-    F: FnOnce() -> Result<Bls12GrothParams>,
-{
-    cache_lookup(&*GROTH_PARAM_MEMORY_CACHE, identifier, generator)
```
Comment (dignifiedquire) [inline filecoin-proofs/src/caches.rs]: Why did you remove the use of the cache? shouldn't we still use that?

**Comment by cryptonemo** [comment]
> @cryptonemo I think you have to update the rust version to make the tests work

I made this PR as an alternative. Any preference?

https://github.com/filecoin-project/merkle_light/pull/68

File: filecoin-proofs/src/caches.rs (hunk 1 above)
Comment (cryptonemo) [reply filecoin-proofs/src/caches.rs]: Compiler was telling me it wasn't used.

Comment (dignifiedquire) [reply filecoin-proofs/src/caches.rs]: I mean, you should keep using it

Comment (cryptonemo) [reply filecoin-proofs/src/caches.rs]: Ok, let me take a look, thanks.

**Comment by cryptonemo** [comment]
> > @cryptonemo I think you have to update the rust version to make the tests work
> 
> I made this PR as an alternative. Any preference?
> 
> [filecoin-project/merkle_light#68](https://github.com/filecoin-project/merkle_light/pull/68)

Could I also get a version bump to make this PR happy?

File: storage-proofs/src/parameter_cache.rs [hunk 2]
```diff
@@ -182,31 +181,14 @@ where
     }
 
     fn get_groth_params<R: RngCore>(
-        rng: Option<&mut R>,
-        circuit: C,
+        _rng: Option<&mut R>,
+        _circuit: C,
```
Comment (dignifiedquire) [inline storage-proofs/src/parameter_cache.rs]: can you remove these argumentes, no need to keep them around

Comment (cryptonemo) [reply storage-proofs/src/parameter_cache.rs]: Hrm, so sort of.  I tried removing both at once and all hell broke loose not knowing about the circuit in other parts, so I reverted for the moment and just removed the rng.

File: fil-proofs-tooling/Cargo.toml [hunk 3]
```diff
@@ -39,7 +39,8 @@ blake2s_simd = "0.5.6"
 fil_logger = "0.1"
 log = "0.4.8"
 uom = "0.26"
-merkletree = "0.14.0"
+#merkletree = "0.15.0"
+merkletree = { git = "https://github.com/filecoin-project/merkle_light" }
```
Comment (dignifiedquire) [inline fil-proofs-tooling/Cargo.toml]: I just release `0.15.1`, so you can switch to that

**Comment by dignifiedquire** [comment]
@cryptonemo looks like this actually broke some tests I am afraid

**Comment by cryptonemo** [comment]
> @cryptonemo looks like this actually broke some tests I am afraid

@dignifiedquire My read of the errors looked like they were v21 param related failures. Not the case?

File: storage-proofs/src/parameter_cache.rs [hunk 4]
```diff
@@ -300,6 +303,7 @@ fn write_cached_verifying_key<E: JubjubEngine>(
     })
 }
 
+#[allow(dead_code)]
 fn write_cached_params<E: JubjubEngine>(
```
Comment (dignifiedquire) [inline storage-proofs/src/parameter_cache.rs]: why is this dead code, isn't this used above Oo

**Comment by dignifiedquire** [comment]
@cryptonemo the test failure looks real

File: storage-proofs/src/parameter_cache.rs (hunk 4 above)
Comment (cryptonemo) [reply storage-proofs/src/parameter_cache.rs]: Yes, was previously unused and I missed it after adding back in, thanks.

File: storage-proofs/src/parameter_cache.rs [hunk 5]
```diff
@@ -181,43 +179,45 @@ where
...
-                Ok(parameters)
-            } else {
-                bail!("No cached parameters found for {}", id);
-            }
+            use rand::SeedableRng;
+            use rand_xorshift::XorShiftRng;
+            use std::time::Instant;
+
+            info!("Actually generating groth params. (id: {})", &id);
+            let rng = &mut XorShiftRng::from_seed(crate::TEST_SEED);
```
Comment (dignifiedquire) [inline storage-proofs/src/parameter_cache.rs]: Sorry I forgot this before, lets not use a seeded rng, but rather `thread_rng`, these should not be regeneratable.

Comment (cryptonemo) [reply storage-proofs/src/parameter_cache.rs]: Ok, thanks, will update this shortly!

**Comment by cryptonemo** [comment]
Updated and rebased!
